Cannot serialize multi-target models when sparse_pickle is True

It is not possible to serialize a model when it is multi-target and sparse_pickle is True

This can be easily tested by updating the `test_serialization` test in `test_quantile_forest.py` and adding a 2d target


```
def check_serialization(name, sparse_pickle, multi_target):
    # Check model serialization/deserialization.

    X = X_california

    if multi_target:
        y = np.vstack([y_california, y_california]).T
    else:
        y = y_california

    ForestRegressor = FOREST_REGRESSORS[name]

    est = ForestRegressor(n_estimators=10, random_state=0)
    est.fit(X, y, sparse_pickle=sparse_pickle)

    dumped = pickle.dumps(est)
    est_loaded = pickle.loads(dumped)

    assert check_is_fitted(est_loaded) is None
    assert np.all(est.predict(X) == est_loaded.predict(X))


@pytest.mark.parametrize("name", FOREST_REGRESSORS)
@pytest.mark.parametrize("sparse_pickle", [False, True])
@pytest.mark.parametrize("multi_target", [False, True])
def test_serialization(name, sparse_pickle, multi_target):
    check_serialization(name, sparse_pickle, multi_target)
```

which yields the following result:

```
============================= test session starts ==============================
collecting ... collected 8 items

test_quantile_forest.py::test_serialization[False-False-ExtraTreesQuantileRegressor] 
test_quantile_forest.py::test_serialization[False-False-RandomForestQuantileRegressor] 
test_quantile_forest.py::test_serialization[False-True-ExtraTreesQuantileRegressor] 
test_quantile_forest.py::test_serialization[False-True-RandomForestQuantileRegressor] 
test_quantile_forest.py::test_serialization[True-False-ExtraTreesQuantileRegressor] 
test_quantile_forest.py::test_serialization[True-False-RandomForestQuantileRegressor] 
test_quantile_forest.py::test_serialization[True-True-ExtraTreesQuantileRegressor] 
test_quantile_forest.py::test_serialization[True-True-RandomForestQuantileRegressor] 

========================= 2 failed, 6 passed in 1.70s ==========================
PASSED [ 12%]PASSED [ 25%]PASSED [ 37%]PASSED [ 50%]PASSED [ 62%]PASSED [ 75%]FAILED [ 87%]
quantile_forest/tests/test_quantile_forest.py:1358 (test_serialization[True-True-ExtraTreesQuantileRegressor])
name = 'ExtraTreesQuantileRegressor', sparse_pickle = True, multi_target = True

    @pytest.mark.parametrize("name", FOREST_REGRESSORS)
    @pytest.mark.parametrize("sparse_pickle", [False, True])
    @pytest.mark.parametrize("multi_target", [False, True])
    def test_serialization(name, sparse_pickle, multi_target):
>       check_serialization(name, sparse_pickle, multi_target)

test_quantile_forest.py:1363: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_quantile_forest.py:1352: in check_serialization
    dumped = pickle.dumps(est)
quantile_forest/_quantile_forest_fast.pyx:536: in quantile_forest._quantile_forest_fast.QuantileForest.__reduce__
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   ValueError: cannot reshape array of size 19940 into shape (2,19940)

quantile_forest/_quantile_forest_fast.pyx:545: ValueError
FAILED [100%]
quantile_forest/tests/test_quantile_forest.py:1358 (test_serialization[True-True-RandomForestQuantileRegressor])
name = 'RandomForestQuantileRegressor', sparse_pickle = True
multi_target = True

    @pytest.mark.parametrize("name", FOREST_REGRESSORS)
    @pytest.mark.parametrize("sparse_pickle", [False, True])
    @pytest.mark.parametrize("multi_target", [False, True])
    def test_serialization(name, sparse_pickle, multi_target):
>       check_serialization(name, sparse_pickle, multi_target)

test_quantile_forest.py:1363: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_quantile_forest.py:1352: in check_serialization
    dumped = pickle.dumps(est)
quantile_forest/_quantile_forest_fast.pyx:536: in quantile_forest._quantile_forest_fast.QuantileForest.__reduce__
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   ValueError: cannot reshape array of size 12900 into shape (2,12900)

quantile_forest/_quantile_forest_fast.pyx:545: ValueError
```

--- a/quantile_forest/_quantile_forest_fast.pyx
+++ b/quantile_forest/_quantile_forest_fast.pyx
@@ -540,7 +540,7 @@ cdef class QuantileForest:
         d = {}
         if self.sparse_pickle:
             matrix1 = kwargs["y_train_leaves"]
-            reshape1 = (matrix1.shape[2], matrix1.shape[0] * matrix1.shape[1] * matrix1.shape[2])
+            reshape1 = (matrix1.shape[3], matrix1.shape[0] * matrix1.shape[1] * matrix1.shape[2])
             d["shape1"] = matrix1.shape
             d["matrix1"] = sparse.csc_matrix(matrix1.reshape(reshape1))
 

--- a/quantile_forest/tests/test_quantile_forest.py
+++ b/quantile_forest/tests/test_quantile_forest.py
@@ -1334,15 +1334,24 @@ def test_monotonic_constraints(name, max_samples_leaf):
     check_monotonic_constraints(name, max_samples_leaf)
 
 
-def check_serialization(name, sparse_pickle):
+def check_serialization(name, sparse_pickle, monotonic_cst, multi_target):
     # Check model serialization/deserialization.
 
     X = X_california
-    y = y_california
+
+    if multi_target:
+        y = np.vstack([y_california, y_california]).T
+    else:
+        y = y_california
+
+    if monotonic_cst and not multi_target:
+        monotonic_cst = [1] * X.shape[1]
+    else:
+        monotonic_cst = None
 
     ForestRegressor = FOREST_REGRESSORS[name]
 
-    est = ForestRegressor(n_estimators=10, random_state=0)
+    est = ForestRegressor(n_estimators=10, monotonic_cst=monotonic_cst, random_state=0)
     est.fit(X, y, sparse_pickle=sparse_pickle)
 
     dumped = pickle.dumps(est)
@@ -1354,8 +1363,10 @@ def check_serialization(name, sparse_pickle):
 
 @pytest.mark.parametrize("name", FOREST_REGRESSORS)
 @pytest.mark.parametrize("sparse_pickle", [False, True])
-def test_serialization(name, sparse_pickle):
-    check_serialization(name, sparse_pickle)
+@pytest.mark.parametrize("monotonic_cst", [False, True])
+@pytest.mark.parametrize("multi_target", [False, True])
+def test_serialization(name, sparse_pickle, monotonic_cst, multi_target):
+    check_serialization(name, sparse_pickle, monotonic_cst, multi_target)
 
 
 def test_calc_quantile():

